vak train crashes when normalize_spectrograms = false

as @yardencsGitHub pointed out in #101 and #110 
(https://github.com/NickleDave/vak/issues/101#issuecomment-596693323)

with the following error
```
Traceback (most recent call last):
  File "/home/nickledave/anaconda3/envs/vak-env/bin/vak", line 11, in <module>
    load_entry_point('vak', 'console_scripts', 'vak')()
  File "/home/nickledave/Documents/repos/vak/src/vak/__main__.py", line 43, in main
    config_file=args.configfile)
  File "/home/nickledave/Documents/repos/vak/src/vak/cli/cli.py", line 21, in cli
    train(toml_path=config_file)
  File "/home/nickledave/Documents/repos/vak/src/vak/cli/train.py", line 114, in train
    [transforms.Lambda(standardize),
  File "/home/nickledave/anaconda3/envs/vak-env/lib/python3.6/site-packages/torchvision/transforms/transforms.py", line 318, in __init__
    assert callable(lambd), repr(type(lambd).__name__) + " object is not callable"
AssertionError: 'NoneType' object is not callable
```

--- a/src/vak/cli/predict.py
+++ b/src/vak/cli/predict.py
@@ -1,4 +1,3 @@
-from functools import partial
 import json
 from pathlib import Path
 
@@ -7,14 +6,13 @@ import joblib
 import pandas as pd
 from tqdm import tqdm
 import torch.utils.data
-from torchvision import transforms
 
 from .. import config
-from ..datasets.unannotated_dataset import UnannotatedDataset
 from .. import io
 from .. import models
-from ..transforms import PadToWindow, ReshapeToWindow
+from .. import transforms
 from .. import util
+from ..datasets.unannotated_dataset import UnannotatedDataset
 
 
 def predict(toml_path):
@@ -40,24 +38,15 @@ def predict(toml_path):
 
     # ---------------- load data for prediction ------------------------------------------------------------------------
     if cfg.predict.spect_scaler_path:
-        standardize = joblib.load(cfg.predict.spect_scaler_path)
+        spect_standardizer = joblib.load(cfg.predict.spect_scaler_path)
     else:
-        standardize = None
+        spect_standardizer = None
 
-    def to_floattensor(ndarray):
-        return torch.from_numpy(ndarray).float()
-
-    # make an "add channel" transform to use with Lambda
-    # this way a spectrogram 'image' has a "channel" dimension (of size 1)
-    # that convolutional layers can work on
-    add_channel = partial(torch.unsqueeze, dim=1)  # add channel at first dimension because windows become batch
-    transform = transforms.Compose(
-        [transforms.Lambda(standardize),
-         PadToWindow(cfg.dataloader.window_size, return_crop_vec=False),
-         ReshapeToWindow(cfg.dataloader.window_size),
-         transforms.Lambda(to_floattensor),
-         transforms.Lambda(add_channel)]
-    )
+    transform, target_transform = transforms.get_defaults('predict',
+                                                          spect_standardizer,
+                                                          window_size=cfg.dataloader.window_size,
+                                                          return_crop_vec=False,
+                                                          )
 
     pred_dataset = UnannotatedDataset.from_csv(csv_path=cfg.predict.csv_path,
                                                split='predict',
@@ -115,7 +104,8 @@ def predict(toml_path):
                                                      num_workers=cfg.predict.num_workers)
 
         # use transform "outside" of Dataset so we can get back crop vec
-        pad_to_window = PadToWindow(cfg.dataloader.window_size, return_crop_vec=True)
+        pad_to_window = transforms.PadToWindow(cfg.dataloader.window_size,
+                                               return_crop_vec=True)
 
         progress_bar = tqdm(data_for_annot)
 

--- a/src/vak/cli/train.py
+++ b/src/vak/cli/train.py
@@ -1,4 +1,3 @@
-from functools import partial
 import json
 import logging
 from pathlib import Path
@@ -9,14 +8,13 @@ from datetime import datetime
 import joblib
 import pandas as pd
 import torch.utils.data
-from torchvision import transforms
 
 from .. import config
 from .. import models
+from .. import transforms
 from .. import util
 from ..datasets.window_dataset import WindowDataset
 from ..io import dataframe
-from ..transforms import StandardizeSpect
 
 
 def train(toml_path):
@@ -77,14 +75,7 @@ def train(toml_path):
     )
 
     # ---------------- load training data  -----------------------------------------------------------------------------
-    if cfg.train.normalize_spectrograms:
-        logger.info('will normalize spectrograms')
-        standardize = StandardizeSpect.fit_df(dataset_df, spect_key=cfg.spect_params.spect_key)
-        joblib.dump(standardize,
-                    results_path.joinpath('StandardizeSpect'))
-    else:
-        standardize = None
-
+    logger.info(f'using training dataset from {cfg.train.csv_path}')
     # below, if we're going to train network to predict unlabeled segments, then
     # we need to include a class for those unlabeled segments in labelmap,
     # the mapping from labelset provided by user to a set of consecutive
@@ -102,25 +93,21 @@ def train(toml_path):
     with open(results_path.joinpath('labelmap.json'), 'w') as f:
         json.dump(labelmap, f)
 
-    logger.info(f'using training dataset from {cfg.train.csv_path}')
-
-    def to_floattensor(ndarray):
-        return torch.from_numpy(ndarray).float()
-    # make an "add channel" transform to use with Lambda
-    # this way a spectrogram 'image' has a "channel" dimension (of size 1)
-    # that convolutional layers can work on
-    add_channel = partial(torch.unsqueeze, dim=0)
-    transform = transforms.Compose(
-        [transforms.Lambda(standardize),
-         transforms.Lambda(to_floattensor),
-         transforms.Lambda(add_channel)]
-    )
-
-    def to_longtensor(ndarray):
-        return torch.from_numpy(ndarray).long()
-    target_transform = transforms.Compose(
-        [transforms.Lambda(to_longtensor)]
-    )
+    # get transforms just before creating datasets with them
+    if cfg.train.normalize_spectrograms:
+        # we instantiate this transform here because we want to save it
+        # and don't want to add more parameters to `transforms.util.get_defaults` function
+        # and make too tight a coupling between this function and that one.
+        # Trade off is that this is pretty verbose (even ignoring my comments)
+        logger.info('will normalize spectrograms')
+        spect_standardizer = transforms.StandardizeSpect.fit_df(dataset_df,
+                                                                spect_key=cfg.spect_params.spect_key)
+        joblib.dump(spect_standardizer,
+                    results_path.joinpath('StandardizeSpect'))
+    else:
+        spect_standardizer = None
+    transform, target_transform = transforms.get_defaults('train',
+                                                          spect_standardizer)
 
     train_dataset = WindowDataset.from_csv(csv_path=cfg.train.csv_path,
                                            split='train',

--- a/src/vak/transforms/__init__.py
+++ b/src/vak/transforms/__init__.py
@@ -1,1 +1,2 @@
 from .transforms import *
+from .util import get_defaults

--- a/src/vak/transforms/functional.py
+++ b/src/vak/transforms/functional.py
@@ -1,4 +1,13 @@
 import numpy as np
+import torch
+
+__all__ = [
+    'pad_to_window',
+    'reshape_to_window',
+    'standardize_spect',
+    'to_floattensor',
+    'to_longtensor',
+]
 
 
 def standardize_spect(spect, mean_freqs, std_freqs, non_zero_std):
@@ -91,3 +100,54 @@ def reshape_to_window(spect, window_size):
     """
     spect_height, spect_width = spect.shape
     return spect.reshape((-1, spect_height, window_size))
+
+
+def to_floattensor(arr):
+    """convert Numpy array to torch.FloatTensor.
+
+    Parameters
+    ----------
+    arr : numpy.ndarray
+
+    Returns
+    -------
+    float_tensor
+        with dtype 'float32'
+    """
+    return torch.from_numpy(arr).float()
+
+
+def to_longtensor(arr):
+    """convert Numpy array to torch.LongTensor.
+
+    Parameters
+    ----------
+    arr : numpy.ndarray
+
+    Returns
+    -------
+    long_tensor : torch.Tensor
+        with dtype 'float64'
+    """
+    return torch.from_numpy(arr).long()
+
+
+def add_channel(input, channel_dim=0):
+    """add a channel dimension to a 2-dimensional tensor.
+    Transform that makes it easy to treat a spectrogram as an image,
+    by adding a dimension with a single 'channel', analogous to grayscale.
+    In this way the tensor can be fed to e.g. convolutional layers.
+
+    Parameters
+    ----------
+    input : torch.Tensor
+        with two dimensions (height, width).
+    channel_dim : int
+        dimension where "channel" is added.
+        Default is 0, which returns a tensor with dimensions (channel, height, width).
+    """
+    if input.dim() != 2:
+        raise ValueError(
+            f'input tensor should have two dimensions but input.dim() is {input.dim()}'
+        )
+    return torch.unsqueeze(input, dim=channel_dim)

--- a/src/vak/transforms/transforms.py
+++ b/src/vak/transforms/transforms.py
@@ -6,7 +6,12 @@ from ..util.validation import column_or_1d
 from . import functional as F
 
 __all__ = [
-    'StandardizeSpect', 'PadToWindow', 'ReshapeToWindow'
+    'AddChannel',
+    'PadToWindow',
+    'ReshapeToWindow',
+    'StandardizeSpect',
+    'ToFloatTensor',
+    'ToLongTensor'
 ]
 
 
@@ -209,3 +214,62 @@ class ReshapeToWindow:
 
     def __call__(self, spect):
         return F.reshape_to_window(spect, self.window_size)
+
+
+class ToFloatTensor:
+    """convert Numpy array to torch.FloatTensor.
+
+    Parameters
+    ----------
+    arr : numpy.ndarray
+
+    Returns
+    -------
+    float_tensor
+        with dtype 'float32'
+    """
+    def __init__(self):
+        pass
+
+    def __call__(self, arr):
+        return F.to_floattensor(arr)
+
+
+class ToLongTensor:
+    """convert Numpy array to torch.LongTensor.
+
+    Parameters
+    ----------
+    arr : numpy.ndarray
+
+    Returns
+    -------
+    long_tensor : torch.Tensor
+        with dtype 'float64'
+    """
+    def __init__(self):
+        pass
+
+    def __call__(self, arr):
+        return F.to_longtensor(arr)
+
+
+class AddChannel:
+    """add a channel dimension to a 2-dimensional tensor.
+    Transform that makes it easy to treat a spectrogram as an image,
+    by adding a dimension with a single 'channel', analogous to grayscale.
+    In this way the tensor can be fed to e.g. convolutional layers.
+
+    Parameters
+    ----------
+    input : torch.Tensor
+        with two dimensions (height, width).
+    channel_dim : int
+        dimension where "channel" is added.
+        Default is 0, which returns a tensor with dimensions (channel, height, width).
+    """
+    def __init__(self, channel_dim=0):
+        self.channel_dim = channel_dim
+
+    def __call__(self, input):
+        return F.add_channel(input, channel_dim=self.channel_dim)

--- a/src/vak/transforms/util.py
+++ b/src/vak/transforms/util.py
@@ -1,0 +1,65 @@
+import torchvision.transforms
+
+from . import transforms as vak_transforms
+
+
+def get_defaults(mode,
+                 spect_standardizer=None,
+                 window_size=None,
+                 return_crop_vec=False,
+                 ):
+    """get default transforms
+
+    Parameters
+    ----------
+    mode : str
+        one of {'train', 'predict'}. Determines set of transforms.
+    spect_standardizer : vak.transforms.StandardizeSpect
+        instance that has already been fit to dataset, using fit_df method.
+        Default is None, in which case no standardization transform is applied.
+
+    Returns
+    -------
+    transform, target_transform : callable
+        one or more vak transforms to be applied to inputs x and, during training, the target y.
+        If more than one transform, they are combined into an instance of torchvision.transforms.Compose.
+        Note that when mode is 'predict', the target transform is None.
+    """
+    # regardless of mode, transform always starts with StandardizeSpect, if used
+    if spect_standardizer is not None:
+        if isinstance(spect_standardizer, vak_transforms.StandardizeSpect):
+            transform = [spect_standardizer]
+        else:
+            raise TypeError(
+                f'invalid type for spect_standardizer: {type(spect_standardizer)}. '
+                'Should be an instance of vak.transforms.StandardizeSpect'
+            )
+    else:
+        transform = []
+
+    if mode == 'train':
+        transform.extend(
+            [
+                vak_transforms.ToFloatTensor(),
+                vak_transforms.AddChannel(),
+            ]
+        )
+
+        target_transform = vak_transforms.ToLongTensor()
+    elif mode == 'predict':
+        transform.extend(
+            [
+                vak_transforms.PadToWindow(window_size, return_crop_vec),
+                vak_transforms.ReshapeToWindow(window_size),
+                vak_transforms.ToFloatTensor(),
+                vak_transforms.AddChannel(channel_dim=1),  # add channel at first dimension because windows become batch
+            ]
+        )
+        target_transform = None
+    else:
+        raise ValueError(
+            f'invalid mode: {mode}'
+        )
+
+    transform = torchvision.transforms.Compose(transform)
+    return transform, target_transform

